src/algorithms.rs

```diff
@@ -109,4 +111,101 @@ impl EGraph {
+    /// a new e-class for each node pointing to any e-class with a node that should be split.
+    ///
+    /// Note that if any of these nodes appear twice in an e-class then it will panic.
+    ///
+    /// Class data will be copied to new nodes.
+    ///
+    /// This can be used for example to make multiple e-classes for all nodes equivalent to i64(0), to make it easier
+    /// to visualize this.
+    pub fn split_e_classes(&mut self, should_split: impl Fn(&NodeId, &Node) -> bool) {
+        // run till fixpoint since splitting a node might add more parents and require splitting the child down the line
+        let mut changed = true;
+        while changed {
+            changed = false;
+            // Mapping from class ID to all nodes that point to any node in that e-class
+            let parents: HashMap<ClassId, Vec<(NodeId, usize)>> =
+                self.nodes
+                    .iter()
+                    .fold(HashMap::new(), |mut parents, (node_id, node)| {
+                        for (position, child) in node.children.iter().enumerate() {
+                            let child_class = self.nodes[child].eclass.clone();
+                            parents
+                                .entry(child_class)
+                                .or_default()
+                                .push((node_id.clone(), position));
+                        }
+                        parents
+                    });
+            for Class { id, nodes } in self.classes().clone().values() {
+                let mut other_nodes = Vec::new();
+                let mut unique_node = None;
+                for node_id in nodes {
+                    let node = self.nodes[node_id].clone();
+                    if should_split(node_id, &node) {
+                        if let Some((other_node_id, other_node)) = unique_node {
+                            panic!(
+                                "Multiple nodes in one e-class should be split. E-class: {:} Node 1: {:?} {:?} Node 2: {:?} {:?}",
+                                id, node_id, node, other_node_id, other_node
+                            );
+                        }
+                        unique_node = Some((node_id, node));
+                    } else {
+                        other_nodes.push(node_id);
+                    }
+                }
+                let class_data = self.class_data.get(id).cloned();
+                if let Some((unique_node_id, unique_node)) = unique_node {
+                    let n_other_nodes = other_nodes.len();
+                    let mut offset = 0;
+                    if n_other_nodes == 0 {
+                        continue;
+                    }
+                    // split out other nodes if there are multiple of them.
+                    // Leave one node in this e-class and make new e-classes for remaining nodes
+                    for other_node_id in other_nodes.into_iter().skip(1) {
+                        changed = true;
```

Comment:
Why is `changed` set to true here (in addition to below where parent E-nodes are updated).

Reply:
I updated the logic to only trigger changed if there are any parents and we make a new node. Changed so only be true if the e-graph was mutated.